### DISCUSSION TITLE
How to get color of dataset?

### DISCUSSION THREAD

**Discussion by AValov76** [GSI-CS-CO/chart-fx#503] [Q&A] [answered]
First, many thanks for great job and really excellent library!

My rather simple question: is it possible to get the color value assigned to the dataset line by a renderer (particulary strokeColor)?
I have some controls related to the chart in which i'd like to show these colors.

**Comment by RalphSteinhagen** [comment, marked as answer]
@AValov76 if you have set them manually via `DataSet::setStyle(...)`, you may want to have a look at the snippets in [DefaultRenderColorScheme](https://github.com/GSI-CS-CO/chart-fx/blob/5ae26b63e35d351630ea83c3b8fab8b41f10dc7e/chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/utils/DefaultRenderColorScheme.java#L235-L247) and retrieve them via for example:

```java
    public static void setLineScheme(final GraphicsContext gc, final String defaultStyle, final int dsIndex) {
        AssertUtils.gtEqThanZero("setLineScheme dsIndex", dsIndex);
        final Map<String, List<String>> map = splitQuery(defaultStyle);


        final Color lineColor = StyleParser.getColorPropertyValue(defaultStyle, XYChartCss.DATASET_STROKE_COLOR);
        final double[] lineDash = StyleParser.getStrokeDashPropertyValue(defaultStyle, XYChartCss.STROKE_DASH_PATTERN);
        final Color rawColor = lineColor == null ? getStrokeColor(dsIndex) : lineColor;


        gc.setLineWidth(defaultStrokeLineWidthProperty().get());
        gc.setLineDashes(lineDash);
        gc.setFill(getFill(dsIndex));
        gc.setStroke(getColorModifier(map, rawColor));
    }
```
Here `defaultStyle` is the style as being set by ` `DataSet::setStyle(...)` and `dsIndex` the counter for the DataSet within a given chart and across the DataSet drawing rendereres attached to the chart. The hatching and fill colors follow a similar syntax.

You may also have a look at how these are being used in the [ErrorDataSetRenderer](https://github.com/GSI-CS-CO/chart-fx/blob/cd202c7c9a0b40f1e66b6d0fe8fd93fffbec5435/chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/ErrorDataSetRenderer.java#L908-L911).

Hope this helps. 

Please share a screenshot of your application. We always love to hear/see how our litte library is being used.

**Comment by AValov76** [comment]
@RalphSteinhagen , thank you for the answer.

It's turned out that I have one renderer per dataset (because I need different DataPointTooltips per graph in a single chart) and I've taken simple approach just getting the color from DefaultRenderColorScheme.TUNEVIEWER by index as it's ok in my case keep track of added/removed datesets.
A couple of screenshots of our app (it's in early development stage now and have not much as planned):

![2022-04-03_18-00-43 (2)](https://user-images.githubusercontent.com/43465649/161436062-de73405d-6783-4144-9151-f19365b6139c.png)
![2022-04-03_17-30-34 (2) (2)](https://user-images.githubusercontent.com/43465649/161436069-a1072bfc-7c55-49c4-a693-26857d226fc3.png)

**Reply by dirri** [reply]
I have similar issue as i need different colors (for line and MARKER)  for different datasets in one chart. Is there a way to have 1 renderer per dataset? As renderes append to the chart an not to datasets...
Currently i have 1 renderer and i overwrite the default Marker as i couldn´t find out how to change ist by configuration.
Now i need at minimum that the second dataset uses a different coloured marker to see the diff in chart.
Is it possible within the renderer to customize the marker for different datasets?

**Reply by ennerf** [reply]
> Is there a way to have 1 renderer per dataset? As renderers append to the chart an not to datasets...

Yes, each renderer can have its own datasets that then show up in the chart, i.e., `renderer.getDataSets().add(dataSet);`. Our company has a few use cases where we need highly customized rendering, and using one renderer per dataset is a pretty common pattern.

The renderers also have a useful toggle for showing/hiding contained datasets from the legend.

**Reply by dirri** [reply]
> > Is there a way to have 1 renderer per dataset? As renderers append to the chart an not to datasets...
> 
> Yes, each renderer can have its own datasets that then show up in the chart, i.e., `renderer.getDataSets().add(dataSet);`. Our company has a few use cases where we need highly customized rendering, and using one renderer per dataset is a pretty common pattern.
> 
> The renderers also have a useful toggle for showing/hiding contained datasets from the legend.

@ennerf  thx a lot the only way i added a renderer before was by adding it to the chart via getRenderers.add ! I´ll gibe it a try and reorganize my project.
It is also very stressful to accomplish as in my case to customize the marker there is no single point of configuration.

**Reply by ennerf** [reply]
For exotic use cases it's sometimes easiest to just write a small custom renderer. The default renderer seems pretty complicated, but that is only because it supports so many different cases. A renderer for drawing markers would only be a few lines of code.

For example, in this use case I use the standard renderer for a line, and then have a second renderer for drawing key points with meta data, e.g., a square for fixed constraints (set by user) or an open circle for free constraints (spline interpolated from other points): https://www.youtube.com/watch?v=kTWuLd8Iacg

**Reply by dirri** [reply]
Again thank you ! 
One last point .. My X-axis is of type Date and i discovered when displaying 2 datasets for comparsion in one chart not from same Date the result is a more or less gap in the chart because the added datasets are from different dates. Is there maybe a way (except rewriting every points in one dataset to same date as the other) to ignore the different 'day' but keep the time scale??  The background for this case are compares of perf measurements and sometimes i want to compare different days performance. Yes ist a special case i know...
btw "The renderers also have a useful toggle for showing/hiding contained datasets from the legend." can you explain this?

**Reply by wirew0rm** [reply]
Either override `getValue` on your dataset (or a custom wrapper dataset) to mask out the date values of your X axis or use the `MathDataSet` to transform the X coordinates (see TSpectrumSample.java) accordingly.
The other logical place for this transform would be the axis, but that makes handling of data that wraps around days more complicated.

re `Renderer::showInLegend(boolena state)`: From javadoc: "Sets whether DataSets attached to this renderer shall be shown in the legend". This is useful if you add your datasets to multiple renderers at the same time. Eg in Florian's example he would set this to false on the second renderer which draws the "key points" to prevent the dataset from being listed in the legend twiice.

**Reply by dirri** [reply]
In fact the x axis contains the timestamp in seconds and these came from a Date object. x axis is a DefaultNumericAxis with timeAxis property set to true, I can filter out the date before adding the timestamp to the DoubleDataSet (as these are Strings received from log files i could use a DateFormatter pattern like (HH:mm:ss) and get the time from the Date object. This works and i can compare 2 series from different dates but the Day that is shown on axis is then set to Jan 01. Additionally this also depends on the zoom level. maybe i need to customize a "TickLabelFormatter" here ?!  The bad is with this solution i loose the Day information.

**Reply by wirew0rm** [reply]
You can definitely change the formatting to not print the date information. For the day you could just add it to the name of the DataSet, so it is shown in the legend?

**Comment by dirri** [comment]
Just to explain what i mean. In the compare case i want to remove the 17-Jul but keep the time and also internally when the 2 charts are displayed the Day should not be included in time axis as then the 2 charts wohnt lie in same area.
![grafik](https://user-images.githubusercontent.com/11440450/179528458-e0dfca21-d34b-4649-af73-65b07c143af3.png)

And what i´ve meant with depends is that the day string disappears when slightly zooming in (ZoomPlugin) : 
![grafik](https://user-images.githubusercontent.com/11440450/179550054-b1bdcd01-2749-4936-bafa-f0c1d8b65efe.png)

So there are 2 two points:
1 comparsion of data from different dates in one area (like fake as they are from same day)
2 When removing the day part from source data (HH:mm:ss) then delete the "01 Jan" String and write correct date into e.g legend where the latter would be the simplest one i think.

**Comment by jdeppe-pivotal** [comment]
I have exactly the same question, but it's 2025 and the classes mentioned here don't exist any more. Could the answer be updated to reflect the current API (`11.3.x`). Thanks!!
